lib/static/new-ui/features/suites/components/SuitesTreeView/utils.ts

```diff
@@ -0,0 +1,389 @@
+                }
+
+                treeNodesCopy[index] = newTreeNode;
+            } else if (treeNode.data.entityType === EntityType.Suite && treeNode.children?.length) {
+                const sortResult = sortAndGetWeight(treeNode.children);
+
+                weights[treeNode.data.id] = extractWeight(treeNode, sortResult.weight);
+
+                const newTreeNode = Object.assign({}, treeNode, {
+                    children: sortResult.sortedTreeNodes
+                });
+
+                const retriesCount = Number(sortResult.weight[0]);
+                if (currentSortExpression?.type === SortType.ByFailedRetries && retriesCount > 0) {
+                    newTreeNode.data.tags.push(`${retriesCount} ${(retriesCount === 1 ? 'failed retry' : 'failed retries')}`);
+                } else if (currentSortExpression.type === SortType.ByTestsCount) {
+                    const testsCount = weights[treeNode.data.id][0] as number;
+                    const retriesCount = weights[treeNode.data.id][1] as number;
+                    newTreeNode.data.tags.push(`${testsCount} ${(testsCount === 1 ? 'test' : ' tests')}`);
+                    newTreeNode.data.tags.push(`${retriesCount} ${(retriesCount === 1 ? 'retry' : ' retries')}`);
+                }
+
+                treeNodesCopy[index] = newTreeNode;
+            } else if (treeNode.data.entityType === EntityType.Browser) {
+                const newTreeNode = Object.assign({}, treeNode);
+
+                weights[treeNode.data.id] = extractWeight(treeNode);
+
+                const retriesCount = weights[treeNode.data.id][0] as number;
+                if (currentSortExpression?.type === SortType.ByFailedRetries && retriesCount > 0) {
+                    newTreeNode.data.tags.push(`${retriesCount} ${(retriesCount === 1 ? 'failed retry' : 'failed retries')}`);
+                }
+
+                treeNodesCopy[index] = newTreeNode;
+            }
+        });
+
+        const sortedTreeNodes = treeNodesCopy.sort((a, b): number => {
+            const direction = currentSortDirection === SortDirection.Desc ? -1 : 1;
+
+            for (let i = 0; i < weights[a.data.id].length; i++) {
+                const aWeight = weights[a.data.id][i];
+                const bWeight = weights[b.data.id][i];
+                if (aWeight === bWeight) {
+                    continue;
+                }
+
+                if (typeof aWeight === 'string' || typeof bWeight === 'string') {
+                    return aWeight.toString().localeCompare(bWeight.toString()) * direction;
+                }
+
+                return (aWeight - bWeight) * direction;
+            }
+
+            return 0;
+        });
+
+        return {
+            sortedTreeNodes: sortedTreeNodes,
+            weight: aggregateWeights(Object.values(weights))
+        };
+    };
+
+    return sortAndGetWeight(treeNodes).sortedTreeNodes;
+};
```

Comment:
a little bit complicated logic here, but i don't how it makes simpler. Just a lot of cases which should be covered.

Reply:
I agree, so I once again looked at the code and tried to simplify it. What's came out of this:
- There is no referring to magic indexes like [0], [1] anymore, I introduced weight metadata instead, which is type-safe and more explicit & readable
- Got rid of duplicating logic related to adding tags, and the code's become a lot cleaner

The file is still pretty large, but more readable and clear.